plan4past/compiler.py

```diff
@@ -174,27 +186,40 @@ def _compile_problem(self):
             else set(self.problem.init)
         )
 
+        goal_predicate = self._predicate_mapping.get_predicate(self.formula)
         self._result_problem = Problem(
             name=self.problem.name,
             domain_name=self.domain.name,
             requirements=self.problem.requirements,
             objects=[*self.problem.objects],
             init=new_init,
-            goal=And(
-                Predicate(add_val_prefix(replace_symbols(to_string(self.formula))))
-            ),
+            goal=And(Predicate(add_val_prefix(goal_predicate.name))),
         )
 
 
-def _compute_whens(formula: Formula) -> Set[When]:
+def _compute_whens(predicates: Set[Predicate], mapping: PredicateMapping) -> Set[When]:
     """Compute conditional effects for formula progression."""
-    return {
-        When(Predicate(add_val_prefix(remove_yesterday_prefix(p.name))), p)
-        for p in predicates(formula)
-    }.union(
-        When(Not(Predicate(add_val_prefix(remove_yesterday_prefix(p.name)))), Not(p))
-        for p in predicates(formula)
-    )
+    result: Set[When] = set()
+    for p in predicates:
+        if p.name.startswith("Y-"):
+            present_predicate = Predicate(
+                add_val_prefix(remove_yesterday_prefix(p.name))
+            )
+            past_predicate = p
+        else:
+            formula = mapping.inverse_mapping[p]
+            if isinstance(formula, Before):
+                arg_predicate = mapping.mapping[formula.argument]
+                present_predicate = Predicate(add_val_prefix(arg_predicate.name))
+                past_predicate = p
+            else:
+                present_predicate = Predicate(add_val_prefix(p.name))
+                past_predicate = p
+
+        positive_when = When(present_predicate, past_predicate)
+        negative_when = When(Not(present_predicate), Not(past_predicate))
+        result.update({positive_when, negative_when})
+    return result
```

Comment:
we cannot just remove the "Y-" prefix since now we have `quoted`. 

However, we still have some predicates as `Y-quoted_0`, which must be handled with the old approach.